### ISSUE TITLE
Add ability to pass raw SQL to onConflict()

### ISSUE THREAD

**Issue by ariwbolton** [knex/knex#4590]
### Use Case

Postgres supports partial unique indices, which unfortunately require some extra work when writing `INSERT ... ON CONFLICT` queries against them:

```sql
INSERT INTO upsertable_table (value)
VALUES ('something')
ON CONFLICT (value) WHERE deleted_at IS NULL
DO NOTHING
RETURNING value, id
```

Note the `WHERE deleted_at IS NULL` portion; that's required so that PG can infer the correct unique index to use for conflict detection.

Unfortunately, `knex` doesn't support this: the `onConflict()` method expects either a `column` or `columns`, and explicitly wraps the formatted columns in parens (as expected): [link to source](https://github.com/knex/knex/blob/bcaf77fcebcbe412e7fa16cf17ba87148e764bab/lib/dialects/postgres/query/pg-querycompiler.js#L78-L88). (So, trying to pass in, say, `knex.raw('(value) WHERE deleted_at IS NULL')` to `onConflict()` results in a broken query, because PG doesn't expect the parens to be wrapped around all of that)

### Feature Requirements/API Proposal

It would be great to be able to use `knex` here even while specifying arbitrarily complex `ON CONFLICT` index expressions, by passing in a raw SQL string, and to have the typings updated to reflect that! (At least, it would be great if this was supported for Postgres; I have no idea if this has any value for other databases)

So, for example, the above query would look something like this, using `knex`:

```javascript
const qb = knex('upsertable_table')
    .insert({ value: 'something' }, ['value', 'id'])
    .onConflict(knex.raw('(value) WHERE deleted_at IS NULL')) // new format!
    .ignore()
```

This proposal is **not** suggesting that `onConflict()` accepts a full query builder; that seems like it might not solve the problem, and might also cause other problems even if it did!

**Comment by avivshafir** [comment]
@newhouse @ariwbolton how did you solve it ? build everything with knex.raw instead?

**Comment by newhouse** [comment]
> @newhouse @ariwbolton how did you solve it ? used `knex.raw` for the insert query instead?

Yes, sadly I built what I needed using `knex.raw`...

**Comment by OlivierCavadenti** [comment]
I implements that, but in integration tests (and manual sql tests) a 'where' like you write dont work in the onConflict process. What is expected behavior (maybe I miss the point) ?
https://stackoverflow.com/a/44579804/1535159

**Comment by newhouse** [comment]
Thanks!

**Comment by kibertoad** [comment]
Released in 1.0.2
